src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,664 @@
+							   relationName)));
+	}
+
+	char relationKind = get_rel_relkind(relationId);
+	if (!(relationKind == RELKIND_RELATION || relationKind == RELKIND_FOREIGN_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", only regular "
+							   "tables and foreign tables are supported for citus local "
+							   "table creation", relationName)));
+	}
+}
+
+
+/*
+ * CreateCitusLocalTable is the internal method that creates a citus table
+ * from the table with relationId. The created table would have the following
+ * properties:
+ *  - it will have only one shard,
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+static void
+CreateCitusLocalTable(Oid relationId)
+{
+	/* these checks should be done before acquiring the lock on the table */
+	EnsureCoordinator();
+	EnsureTableOwner(relationId);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relationId);
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	List *shellTableDDLEvents = GetShellTableDDLEventsForCitusLocalTable(relationId);
+
+	char *relationName = get_rel_name(relationId);
+	Oid relationSchemaId = get_rel_namespace(relationId);
+
+	/* below we convert relation with relationId to the shard relation */
+	uint64 shardId = ConvertLocalTableToShard(relationId);
+
+	/*
+	 * As we retrieved the DDL commands necessary to create the shell table
+	 * from scratch, below we simply recreate the shell table executing them
+	 * via process utility.
+	 */
+	CreateShellTableForCitusLocalTable(shellTableDDLEvents);
+
+	/*
+	 * Get new relationId as the relation with relationId now points
+	 * to the shard relation.
+	 */
+	Oid shellRelationId = get_relname_relid(relationName, relationSchemaId);
+
+	/* assert that we created the shell table properly in the same schema */
+	Assert(OidIsValid(shellRelationId));
+
+	InsertMetadataForCitusLocalTable(shellRelationId, shardId);
+
+	FinalizeCitusLocalTableCreation(relationId);
+
+	relation_close(relation, ExclusiveLock);
```

Comment:
Why are we releasing the lock? I think we should close with `NoLock`, keep the lock until the end of the transaction. (Still, adding this to discussion with Marco)

Reply:
Now as we pushed all the operations done relation` into that function (`CreateCitusLocalTable`), isn't it good to release the lock at the end ?

Reply:
I think Talha's test reveals the issue. Holding the lock until the transaction ends is necessary

> open two backends
create a table 'create table local(a int);' in the first one
run select create_citus_local_table('local') but stop it right after ConvertLocalTableToShard is executed, so at this point the postgres table is converted to a shard but shell table(empty table) is not created yet.
in the other backend run 'select create_citus_local_table('local');', now this will block in Relation relation = relation_open(relationId, ExclusiveLock); at CreateCitusLocalTable
Continue in the first one, so it creates the citus local table
Now as the lock is released from the first backend, the second one continues and it doesn't fail and in pg_dist_partition we have an extra entry now:

Reply:
I will try and update here

Reply:
As I also tried locally, " not releasing the lock" doesn't help to solve the problem that the @SaitTalhaNisanci mentioned. This because, either we release the lock at the end of xact or when the all the creation logic is done in  `CreateCitusLocalTable`, the other backend would still try to create a citus local table from the `relationId` and this is indeed valid as that relation is still not a citus relation, but it became a shard relation.

So, I purpose to check for the the `relationId` with `RelationIsAKnownShard` function to error out.

Reply:
Does it sound good ?